tools/wpt.ts

```diff
@@ -0,0 +1,532 @@
+}
+
+interface TestToRun {
+  sourcePath: string;
+  path: string;
+  url: URL;
+  options: ManifestTestOptions;
+  expectation: boolean | string[];
+}
+
+async function run() {
+  assert(Array.isArray(rest), "filter must be array");
+  const tests = discoverTestsToRun(rest.length == 0 ? undefined : rest);
+  console.log(`Going to run ${tests.length} test files.`);
+
+  const results = await runWithTestUtil(false, async () => {
+    const results = [];
+
+    for (const test of tests) {
+      console.log(`${blue("-".repeat(40))}\n${bold(test.path)}\n`);
+      const result = await runSingleTest(
+        test.url,
+        test.options,
+        json ? () => {} : createReportTestCase(test.expectation),
+      );
+      results.push({ test, result });
+      reportVariation(result, test.expectation);
+    }
+
+    return results;
+  });
+
+  if (json) {
+    await Deno.writeTextFile(json, JSON.stringify(results));
+  }
+  const code = reportFinal(results);
+  Deno.exit(code);
+}
+
+async function update() {
+  assert(Array.isArray(rest), "filter must be array");
+  const tests = discoverTestsToRun(rest.length == 0 ? undefined : rest, true);
+  console.log(`Going to run ${tests.length} test files.`);
+
+  const results = await runWithTestUtil(false, async () => {
+    const results = [];
+
+    for (const test of tests) {
+      console.log(`${blue("-".repeat(40))}\n${bold(test.path)}\n`);
+      const result = await runSingleTest(
+        test.url,
+        test.options,
+        json ? () => {} : createReportTestCase(test.expectation),
+      );
+      results.push({ test, result });
+      reportVariation(result, test.expectation);
+    }
+
+    return results;
+  });
+
+  if (json) {
+    await Deno.writeTextFile(json, JSON.stringify(results));
+  }
+
+  const resultTests: Record<
+    string,
+    { passed: string[]; failed: string[]; status: number }
+  > = {};
+  for (const { test, result } of results) {
+    if (!resultTests[test.sourcePath]) {
+      resultTests[test.sourcePath] = {
+        passed: [],
+        failed: [],
+        status: result.status,
+      };
+    }
+    for (const case_ of result.cases) {
+      if (case_.passed) {
+        resultTests[test.sourcePath].passed.push(case_.name);
+      } else {
+        resultTests[test.sourcePath].failed.push(case_.name);
+      }
+    }
+  }
+
+  const currentExpectation = getExpectation();
+
+  for (const path in resultTests) {
+    const { passed, failed, status } = resultTests[path];
+    let finalExpectation: boolean | string[];
+    if (failed.length == 0 && status == 0) {
+      finalExpectation = true;
+    } else if (failed.length > 0 && passed.length > 0 && status == 0) {
+      finalExpectation = failed;
+    } else {
+      finalExpectation = false;
+    }
+
+    insertExpectation(
+      path.slice(1).split("/"),
+      currentExpectation,
+      finalExpectation,
+    );
+  }
+
+  await Deno.writeTextFile(
+    "./tools/wpt/expectation.json",
```

Comment:
Resolve relative to this file? Otherwise it will fail if not run from root of the repo